typo: fix missing backtick (#20202)
cloudflare/cloudflare-docs: src/content/docs/agents/examples/schedule-tasks.mdx

---
title: Schedule tasks
pcx_content_type: concept
sidebar:
  order: 5
---

import { MetaInfo, Render, Type, TypeScriptExample, WranglerConfig } from "~/components";

> An Agent can schedule tasks to be run in the future by calling `this.schedule(when, callback, data)`, where `when` can be a delay, a `Date`, or a cron string; `callback` the function name to call, and `data` is an object of data to pass to the function.

Reviewer: ```suggestion
An Agent can schedule tasks to be run in the future by calling `this.schedule(when, callback, data)`, where:
  - `when` can be a delay, a `Date`, or a cron string
  - `callback` is the function name to call
  - `data` is an object of data to pass to the function
```

> Scheduled tasks can do anything a request or message from a user can: make requests, query databases, send emails, read+write state: scheduled tasks can invoke any regular method on your Agent.

Reviewer: ```suggestion
Scheduled tasks can do anything a request or message from a user can: make requests, query databases, send emails, read/write state: scheduled tasks can invoke any regular method on your Agent.
```

	async onRequest(request) {
		// Handle an incoming request
		// Schedule a task 5 minutes from now
		// Calls the "checkFlights" method
		let { taskId } = await this.schedule(600, "checkFlights", { flight: "DL264", date: "2025-02-23" });
		return Response.json({ taskId });
	}

	async checkFlights(data) {
		// Invoked when our scheduled task runs
		// We can also call this.schedule here to schedule another task
	}
}
```
</TypeScriptExample>

You can schedule tasks in multiple ways:

<TypeScriptExample>

```ts
// schedule a task to run in 10 seconds
let task = await this.schedule(10, "someTask", { message: "hello" });

// schedule a task to run at a specific date
let task = await this.schedule(new Date("2025-01-01"), "someTask", {});

// schedule a task to run every 10 seconds
let { id } = await this.schedule("*/10 * * * *", "someTask", { message: "hello" });

// schedule a task to run every 10 seconds, but only on Mondays
let task = await this.schedule("0 0 * * 1", "someTask", { message: "hello" });

// cancel a scheduled task
this.cancelSchedule(task.id);
```

</TypeScriptExample>

:::caution

Tasks that set a callback for a method that does not exist will fail silently.

:::

> Calling `await this.schedule` returns a `Schedule`, which includes the task's randomly generated `id`. You can use this `id` to retrieve or cancel the task in the future. It also provides a `type` property that indicates the type of schedule - e.g. one of `"scheduled" | "delayed" | "cron"`.

Reviewer: ```suggestion
Calling `await this.schedule` returns a `Schedule`, which includes the task's randomly generated `id`. You can use this `id` to retrieve or cancel the task in the future. It also provides a `type` property that indicates the type of schedule, for example, one of `"scheduled" | "delayed" | "cron"`.
```

:::note[Maximum scheduled tasks]

> Each task is mapped to a row in the Agent's underlying [SQLite database](/sqlite-in-durable-objects/), which means that each task can be up to 2MB in size. The maximum number of tasks must be `(task_size * tasks) + all_other_state < maximum_database_size` (currently 1GB per Agent).

Reviewer: ```suggestion
Each task is mapped to a row in the Agent's underlying [SQLite database](/sqlite-in-durable-objects/), which means that each task can be up to 2 MB in size. The maximum number of tasks must be `(task_size * tasks) + all_other_state < maximum_database_size` (currently 1GB per Agent).
```

:::

### Managing scheduled tasks

> You can get, cancel and filter across scheduled tasks within an Agent using the scheduling API:

Reviewer: ```suggestion
You can get, cancel, and filter across scheduled tasks within an Agent using the scheduling API:
```
